Applications/CausalLM/layers/mha_core.cpp

```diff
@@ -0,0 +1,767 @@
+
+  ml::train::TensorDim query_dim =
+    query.getDim(); // (B, 1, seq_len, n_heads_Q * head_dim)
+  ml::train::TensorDim key_dim =
+    key.getDim(); // (B, 1, seq_len, n_heads_KV * head_dim)
+  ml::train::TensorDim value_dim =
+    value.getDim(); // (B, 1, seq_len, n_heads_KV * head_dim)
+  ml::train::TensorDim output_dim =
+    output.getDim(); // (B, 1, seq_len, n_heads_Q * head_dim)
+  ml::train::TensorDim cache_key_dim =
+    cache_key.getDim(); // (B, 1, max_seq_len, n_heads_KV * head_dim)
+  ml::train::TensorDim cache_value_dim =
+    cache_value.getDim(); // (B, 1, max_seq_len, n_heads_KV * head_dim)
+
+  ml::train::TensorDim query_step_dim =
+    get_step_dim(query_dim); // (B, 1, from-to, n_heads_Q * head_dim)
+  ml::train::TensorDim output_step_dim =
+    get_step_dim(output_dim); // (B, 1, from-to, n_heads_Q * head_dim)
+  ml::train::TensorDim cache_key_step_dim =
+    get_step_dim(cache_key_dim); // (B, 1, from-to, n_heads_KV * head_dim)
+
+  ml::train::TensorDim cache_value_step_dim =
+    get_step_dim(cache_value_dim); // (B, 1, from-to, n_heads_KV * head_dim)
+
+  unsigned int batch_size = (_from) ? 1 : query_dim.batch();
+  // auto start_time = std::chrono::high_resolution_clock::now();
+  // do the incremental forwarding
+  for (unsigned int batch = 0; batch < batch_size; ++batch) {
+    one_batch_incremental_forwarding(
+      batch, _from, from, to, query, key, value, output, cache_key, cache_value,
+      query_dim, query_step_dim, key_dim, value_dim, cache_key_dim,
+      cache_key_step_dim, cache_value_dim, cache_value_step_dim, output_dim,
+      output_step_dim);
+  }
+
+  if (!_from) {
+    batch_size = query_dim.batch();
+    nntrainer::Tensor cache_key_0_step =
+      cache_key.getSharedDataTensor(cache_key_step_dim, 0, true);
+    nntrainer::Tensor cache_value_0_step =
+      cache_value.getSharedDataTensor(cache_value_step_dim, 0, true);
+
+    for (unsigned int batch = 1; batch < batch_size; ++batch) {
+      nntrainer::Tensor cache_key_nth_step = cache_key.getSharedDataTensor(
+        cache_key_step_dim,
+        batch * cache_key_dim.getFeatureLen() + from * cache_key_dim.width(),
+        true);
+      nntrainer::Tensor cache_value_nth_step =
+        cache_key.getSharedDataTensor(cache_value_step_dim,
+                                      batch * cache_value_dim.getFeatureLen() +
+                                        from * cache_value_dim.width(),
+                                      true);
+
+      cache_key_nth_step.copyData(cache_key_0_step);
+      cache_key_nth_step.copyData(cache_value_0_step);
+    }
+  }
+}
+
+void MHACoreLayer::compute_kcaches(
+  nntrainer::Tensor &in, nntrainer::Tensor &cache, nntrainer::Tensor &out,
+  unsigned int from, size_t sequence_len, unsigned int num_head,
+  unsigned int group_size, unsigned int head_dim, BS::thread_pool<> &pool) {
+
+  if (in.getDataType() == ml::train::TensorDim::DataType::FP32) {
+    if (from) {
+      nntrainer::compute_kcaches<uint16_t>(
```

Comment:
Here, we are not using Multi-Threading for Token Generation case. So for the better performance, we do need optimize this ```nntrainer::computekcaches``` function with multi-threading.